generic: constness: Fix eq_literal

IagoAbal opened the pull request:
The previous definition could cause an infinite loop, e.g.:

    def run():
        while c:
            response = None

Here the old `eq_literal` would say that `None` and `None` were
different, hence we never reached a fixpoint.

IagoAbal commented:
The CI error doesn't seem related to my commit...
> 2021-01-08T21:26:36.0362496Z == [ERROR] you didn't configure what rules semgrep should scan for.

aryx commented on lang_GENERIC/analyze/Dataflow_constness.ml:
```diff
@@ -62,16 +62,7 @@ let str_of_name ((s, _tok), sid) =
-  | G.String (s1, _), G.String (s2, _) -> s1 =$= s2
-  (* add more cases if needed *)
-  | ___else___ -> false
+let eq_literal l1 l2 = G.equal_any (G.E (G.L l1)) (G.E (G.L l2))
```
You can use G.equal_literal I think directly.

aryx replied:
the [@@deriving] in AST_generic.ml should generate some show_xxx and equal_xx for all the types.

aryx commented:
I don't understand why there was an infinute loop.
None was parsed as a literal? And then the __else__ case was taken, but why we never reached a fixpoint?

aryx commented:
Also when we have such infinite loop problem, it would be great to add an actual test file somewhere to actually exercise the new behavior.
Maybe we can have a direcory in pfff/tests/python/constness/ with many test file on which we run the dataflow analysis. See Unit_na,ing_generic.ml for example of testing code (but more generall all the Unit_xxx.ml).

IagoAbal commented:
>  None was parsed as a literal?

Yes, `None` is parsed as the literal `Null` it seems.

> And then the else case was taken, but why we never reached a fixpoint?

The fixpoint is reached when the input and output mappings of an iteration become equal (each variable is mapped to the same `constness` in both the input and output mappings). But this never happens because for the mappings to be equal, `Null` and `Null` should be considered equal, and they are considered to be different.

IagoAbal committed:
generic: constness: Fix eq_literal

The previous definition could cause an infinite loop, e.g.:

    def run():
        while c:
            response = None

Here the old `eq_literal` would say that `None` and `None` were
different, hence we never reached a fixpoint.

aryx commented:
Ah ok. Got it. 

Sent from my iPhone

> On 10 Jan 2021, at 18:33, Iago Abal <notifications@github.com> wrote:
> 
> ﻿
> None was parsed as a literal?
> 
> Yes, None is parsed as the literal Null it seems.
> 
> And then the else case was taken, but why we never reached a fixpoint?
> 
> The fixpoint is reached when the input and output mappings of an iteration become equal (each variable is mapped to the same constness in both the input and output mappings). But this never happens because for the mappings to be equal, Null and Null should be considered equal, and they are considered to be different.
> 
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub, or unsubscribe.

aryx reviewed (approved):
This is perfect! Thx for adding the test.
